### PR TITLE
Api deprecate cmap functions

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

This is my attempt to turn the crank on the depractions on the cm functions.

The changes were more extensive than I anticipated 

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] New features are documented, with examples if plot related.
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: doc/api/next_api_changes/deprecations/XYZ-TC.rst [hunk 1]
```diff
@@ -0,0 +1,14 @@
+Deprecated top-level cmap registration and access functions in ``mpl.cm``
```
Comment (greglucas) [inline doc/api/next_api_changes/deprecations/XYZ-TC.rst:1]:
```suggestion
Pending deprecation of top-level cmap registration and access functions in ``mpl.cm``
```

File: doc/api/next_api_changes/deprecations/XYZ-TC.rst [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As part of `multi-step process
+<https://github.com/matplotlib/matplotlib/issues/20853>`_ we are refactoring
+global state for managing the registered colormap.
```
Comment (greglucas) [inline doc/api/next_api_changes/deprecations/XYZ-TC.rst:6]:
```suggestion
As part of a `multi-step process
<https://github.com/matplotlib/matplotlib/issues/20853>`_ we are refactoring
the global state for managing the registered colormaps.
```

File: examples/lines_bars_and_markers/multivariate_marker_plot.py [hunk 3]
```diff
@@ -31,7 +31,7 @@
 positions = np.random.normal(size=(N, 2)) * 5
 data = zip(skills, takeoff_angles, thrusts, successful, positions)
 
-cmap = plt.cm.get_cmap("plasma")
```
Comment (greglucas) [inline examples/lines_bars_and_markers/multivariate_marker_plot.py:34]: I didn't realize there was also cm access through pyplot via `plt.cm.get_cmap`!

File: lib/matplotlib/cm.py [hunk 4]
```diff
@@ -201,6 +201,11 @@ def unregister(self, name):
 globals().update(_colormaps)
 
 
+@_api.deprecated(
+    '3.6',
+    pending=True,
+    alternative="``matplotilb.colormaps.unregister_cmap(name)``"
```
Comment (greglucas) [inline lib/matplotlib/cm.py:207]:
```suggestion
    alternative="``matplotilb.colormaps.register_cmap(name)``"
```

File: lib/matplotlib/cm.py [hunk 5]
```diff
@@ -663,3 +678,11 @@ def changed(self):
     *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*
     name together with *vmin*/*vmax* is acceptable).""",
 )
+
+
+def _normalize_cmap(cmap):
```
Comment (greglucas) [inline lib/matplotlib/cm.py:683]: I was a bit confused with the name "normalize" here at first. I'm not sure I have a better one, but when you're also dealing with "resampling" in the PR it threw me off. Maybe `_get_cmap_or_default`?

**Review by greglucas — COMMENTED** [review]
I will say, this looks like it will be noisy, but I think that it all makes everything much more explicit and readable IMO. So, overall I think it is a good direction.

File: lib/matplotlib/cm.py (hunk 5 above)
Comment (tacaswell) [reply lib/matplotlib/cm.py]: I tried `_ensure_cmap` because it will also short-circuit on being passed a `Colormap` object.

**Comment by tacaswell** [comment]
OK, really done force-pushing now...

**Review by greglucas — APPROVED** [review]
:+1: Good choice with the name of the private function, I think that is better.

File: lib/matplotlib/cm.py [hunk 6]
```diff
@@ -663,3 +678,26 @@ def changed(self):
...
+def _ensure_cmap(cmap):
+    """
+    Ensure that we have a `.Colormap` object.
+
+    Parameters
+    ----------
+    cmap : None, str, Colormap
+
+        - if a `Colormap` return it
+        - in a string look it up in mpl.colormaps
```
Comment (QuLogic) [inline lib/matplotlib/cm.py:692]:
```suggestion
        - if a `Colormap`, return it
        - if a string, look it up in mpl.colormaps
```

File: lib/matplotlib/pyplot.py [hunk 7]
```diff
@@ -2084,7 +2084,7 @@ def set_cmap(cmap):
     matplotlib.cm.register_cmap
     matplotlib.cm.get_cmap
     """
-    cmap = cm.get_cmap(cmap)
+    cmap = colormaps[cmap]
```
Comment (QuLogic) [inline lib/matplotlib/pyplot.py:2087]: Should this not use `_ensure_cmap` to accept a `Colormap` instance?

File: lib/matplotlib/tests/test_axes.py [hunk 8]
```diff
@@ -2250,7 +2250,7 @@ def test_contour_hatching():
     x, y, z = contour_dat()
     fig, ax = plt.subplots()
     ax.contourf(x, y, z, 7, hatches=['/', '\\', '//', '-'],
-                cmap=plt.get_cmap('gray'),
+                cmap=mpl.colormaps['gray'],
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py:2253]: Can we just use strings if we aren't using `resample` or using the `Colormap` for anything?

File: lib/matplotlib/tests/test_colors.py [hunk 9]
```diff
@@ -64,44 +65,53 @@ def test_resample():
 
 
 def test_register_cmap():
-    new_cm = cm.get_cmap("viridis")
+    new_cm = mpl.colormaps["viridis"]
     target = "viridis2"
-    cm.register_cmap(target, new_cm)
-    assert plt.get_cmap(target) == new_cm
+    with pytest.warns():
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_colors.py:70]: Can we add matching text/type on the `warns`?

File: tutorials/colors/colormap-manipulation.py [hunk 10]
```diff
@@ -25,18 +25,18 @@
 
 First, getting a named colormap, most of which are listed in
 :doc:`/tutorials/colors/colormaps`, may be done using
-`.matplotlib.cm.get_cmap`, which returns a colormap object.
+`.matplotlib.colormaps`, which returns a colormap object.
 The second argument gives the size of the list of colors used to define the
 colormap, and below we use a modest value of 8 so there are not a lot of
 values to look at.
```
Comment (QuLogic) [inline tutorials/colors/colormap-manipulation.py:31]: There's no second argument to a dictionary access. Should change to talk about `resample`.

File: lib/matplotlib/pyplot.py (hunk 7 above)
Comment (tacaswell) [reply lib/matplotlib/pyplot.py:2087]: It should.

File: lib/matplotlib/tests/test_axes.py (hunk 8 above)
Comment (tacaswell) [reply lib/matplotlib/tests/test_axes.py:2253]: It can, but I was just mechanically changing (a lot of this was done with search-and-relpace)

**Comment by tacaswell** [comment]
I'll try to get these changes in first thing tomorow morning, will not be offended if you want to push change @QuLogic .

File: doc/api/next_api_changes/deprecations/23668-TC.rst [hunk 11]
```diff
@@ -0,0 +1,14 @@
+Pending deprecation top-level cmap registration and access functions in ``mpl.cm``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As part of a `multi-step process
+<https://github.com/matplotlib/matplotlib/issues/20853>`_ we are refactoring
+the global state for managing the registered colormaps.
+
+In Matplotlib 3.5 we added a `.ColormapRegistry` class and exposed an
+instance at the top level as ``matplotlib.colormaps``.  The existing
+top level functions in `matplotlib.cm` (``get_cmap``, ``register_cmap``,
```
Comment (timhoffm) [inline doc/api/next_api_changes/deprecations/23668-TC.rst:10]: I just realized that `get_cmap` and `register_cmap` are imported into `pyplot`, which makes the deprecation active also there.

https://github.com/matplotlib/matplotlib/blob/2b96795e324db95509ed1f669c471919685056d6/lib/matplotlib/pyplot.py#L71

IMHO it's ok to deprecate `register_cmap` in `pyplot`. However, I think that `plt.get_cmap()` is too widely used to be deprecated. Therefore we should implement a `plt.get_cmap()` adapter in `pyplot` (maybe deprecate the `lut` parameter there in the future). If that's too much for 3.6, it's ok to only add the pending deprecation to `register_cmap` and `unregister_cmap` and handle `get_cmap` in 3.7.

File: lib/matplotlib/colors.py [hunk 12]
```diff
@@ -854,7 +854,7 @@ def is_gray(self):
         return (np.all(self._lut[:, 0] == self._lut[:, 1]) and
                 np.all(self._lut[:, 0] == self._lut[:, 2]))
 
-    def _resample(self, lutsize):
+    def resample(self, lutsize):
```
Comment (timhoffm) [inline lib/matplotlib/colors.py:857]: I'm tempted to name this `resampled()`, which IMHO makes it more clear that it returns a new colormap and does not modify in place. This is also consistent with `Colormap.reversed()`.

Comment (tacaswell) [reply lib/matplotlib/colors.py:857]: I agree on this.  Should also help make clear this is returning a new color map not mutating the one we have.

File: doc/api/next_api_changes/deprecations/23668-TC.rst (hunk 11 above)
Comment (tacaswell) [reply doc/api/next_api_changes/deprecations/23668-TC.rst:10]: So the tasks here are:

 - implement a non-warning `get_cmap` in pyplot
 - track down everyplace `register_cmap` and `unregister_cmap` are imported and make sure their deprecation is noted
 - update deprecation note to note all of the above

File: lib/matplotlib/pyplot.py (hunk 7 above)
Comment (tacaswell) [reply lib/matplotlib/pyplot.py:2087]: sigh, there are exception stability issues, this used to raise `ValueError` but `_ensure_cmap` will raise `KeyError`.

File: lib/matplotlib/tests/test_axes.py (hunk 8 above)
Comment (tacaswell) [reply lib/matplotlib/tests/test_axes.py:2253]: I've decided to leave this.  It makes the test a bit more complicated, but I do not know if we are definitely exercising the ability to pass `Colormap` objects through this top level API else where.

**Comment by tacaswell** [comment]
@timhoffm I think this is ready to review again.

File: doc/api/next_api_changes/deprecations/23668-TC.rst (hunk 11 above)
Comment (tacaswell) [reply doc/api/next_api_changes/deprecations/23668-TC.rst:10]: I believe this is done.

File: tutorials/colors/colormap-manipulation.py [hunk 13]
```diff
@@ -24,19 +24,19 @@
...
 
 First, getting a named colormap, most of which are listed in
-:doc:`/tutorials/colors/colormaps`, may be done using
-`.matplotlib.cm.get_cmap`, which returns a colormap object.
-The second argument gives the size of the list of colors used to define the
-colormap, and below we use a modest value of 8 so there are not a lot of
-values to look at.
+:doc:`/tutorials/colors/colormaps`, may be done using `.matplotlib.colormaps`,
+which returns a colormap object.  The length of the list of colors used
+internally to define the colormap and be adjusted via `.Colormap.resampled`.
```
Comment (timhoffm) [inline tutorials/colors/colormap-manipulation.py:29]:
```suggestion
which returns a colormap object.  The length of the list of colors used
internally to define the colormap can be adjusted via `.Colormap.resampled`.
```

File: tutorials/colors/colormap-manipulation.py [hunk 14]
```diff
@@ -24,19 +24,19 @@
...
+Blow we use a modest value of 8 so there are not a lot of values to look at.
```
Comment (timhoffm) [inline tutorials/colors/colormap-manipulation.py:30]:
```suggestion
Below we use a modest value of 8 so there are not a lot of values to look at.
```

File: lib/matplotlib/pyplot.py [hunk 15]
```diff
@@ -2069,6 +2069,11 @@ def clim(vmin=None, vmax=None):
     im.set_clim(vmin, vmax)
 
 
+def get_cmap(name=None, lut=None):
+    return cm._get_cmap(name=name, lut=lut)
+get_cmap.__doc__ = cm._get_cmap.__doc__
```
[the remark this answers is not in the document]
Comment (tacaswell) [reply lib/matplotlib/pyplot.py:2074]: I also went back and forth a few times and settled in the indirection.  This way we are sure there are no expected API changes from trying to "improve" the code.

File: lib/matplotlib/tests/test_colors.py [hunk 16]
```diff
@@ -64,44 +65,71 @@ def test_resample():
...
+            match=r"matplotilb.colormaps.unregister_cmap\(name\)"
+    ):
+        # test that second time is error free
+        cm.unregister_cmap(target)
 
     with pytest.raises(TypeError, match="'cmap' must be"):
-        cm.register_cmap('nome', cmap='not a cmap')
+        with pytest.warns(
+            PendingDeprecationWarning,
+            match=r"matplotilb.colormaps.register_cmap\(name\)"
```
[the remark this answers is not in the document]
Comment (tacaswell) [reply lib/matplotlib/tests/test_colors.py:107]: I was fighting with "wont this match at all" and eventually got to only needing to escape the `(` for it to pass.  I'll put escaping on all of them.

File: lib/matplotlib/tests/test_colors.py [hunk 17]
```diff
@@ -64,44 +65,71 @@ def test_resample():
...
-    cm.unregister_cmap(target)
+        with pytest.warns(
+            PendingDeprecationWarning,
+            match=r"matplotilb\.colormaps\.register_cmap\(name\)"
+        ):
+            cm.register_cmap()
+
+    with pytest.warns(
+            PendingDeprecationWarning,
+            match=r"matplotilb.colormaps.unregister_cmap\(name\)"
```
[the remark this answers is not in the document]
Comment (tacaswell) [reply lib/matplotlib/tests/test_colors.py:87]: I suspect this also says I have a misspelling in the code...

File: lib/matplotlib/pyplot.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (tacaswell) [reply lib/matplotlib/pyplot.py:2076]: Done and added `pyplot.get_cmap` to the pyplot documentation (and removed the cut-out above).

**Comment by tacaswell** [comment]
@timhoffm I think I got everything

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
Anybody can merge when green.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**